JsonSubTypes/JsonSubtypes.cs

```diff
@@ -444,9 +453,30 @@ private static object ThreadStaticReadObject(JsonReader reader, JsonSerializer s
             }
         }
 
+        private static IEnumerable<object> GetAttributes(TypeInfo typeInfo)
+        {
+#if NET35
+            if (_attributesCache.TryGetValue(typeInfo, out var res))
+                return res;
+
+            lock (_attributesCache)
+            {
+                if (!_attributesCache.TryGetValue(typeInfo, out res))
+                {
+                    res = typeInfo.GetCustomAttributes(false);
+                    _attributesCache.Add(typeInfo, res);
+                }
+            }
+
+            return res;
+#else
+            return _attributesCache.GetOrAdd(typeInfo, typeInfo.GetCustomAttributes(false));
```

Comment:
Unfortunately you've lost all your optimization since `typeInfo.GetCustomAttributes(false)` is evaluated each call now. Move it into separate static method and use GetOrAdd overload with delegate.

Reply:
Oh, sorry. Used `Func<>` to fix this. Or, is separate method more appropriate in this case?

Reply:
Separate static Func<> will be 'safer' because roslyn (in fact due to ECMA spec requirements) has some issues when creating delegates from method groups or lambdas.

Reply:
More info here https://github.com/dotnet/roslyn/issues/39869

Reply:
Used separate static Func<>